### ISSUE TITLE
Support linux binary for arm architecture

### ISSUE THREAD

**Issue by ryuheechul** [sourcegraph/sg.nvim#154]
## Motivation

Currently, the failure message for arm linux platform is similar to https://github.com/sourcegraph/sg.nvim/issues/113.
Which is not specific and require digging to figure out what went wrong.

## Use case
- virtual machines on m1 macbook 
- as well as native arm architecture linux machines

## Potential fixes
- simply supporting the arm binary should resolve the issue
- or could fail more clearly

## Files requires changes

few additional lines can be added to the files below to produce the binary and select:
- https://github.com/sourcegraph/sg.nvim/blob/acdc88fb410762828797e8816fa790b96b71bb97/lua/sg/build.lua#L24-L26
- https://github.com/sourcegraph/sg.nvim/blob/acdc88fb410762828797e8816fa790b96b71bb97/.github/workflows/release.yml#L92

**Comment by tjdevries** [comment]
What is the output of: `:lua =vim.loop.os_uname()` on your machine?

**Comment by ryuheechul** [comment]
```lua
{
  machine = "aarch64",
  release = "6.1.45",
  sysname = "Linux",
  version = "#1-NixOS SMP Fri Aug 11 10:08:27 UTC 2023"
}
```

**Comment by tjdevries** [comment]
Ok, I did my best on this without something to test on easily (i can try and dig something up if we run into some trouble).

Hopefully with latest release and binaries this works? Could you give it a try for me?

**Comment by ryuheechul** [comment]
Sure and this is the result!

```bash
$ file ~/.local/share/nvim/lazy/sg.nvim/dist/sg-nvim-agent
sg-nvim-agent: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=3b6b0fa621381e5f600e88f1e4cb50218e5879ab, for GNU/Linux 3.7.0, with debug_info, not stripped
$ file ~/.local/share/nvim/lazy/sg.nvim/dist/sg-lsp
sg-lsp: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=fe5594af58a599f9f1308d22737b7a6effa52335, for GNU/Linux 3.7.0, with debug_info, not stripped
```

Which indicates that it should work with most "regular" linux distros as long as they have the path and the file ready for `/lib/ld-linux-aarch64.so.1`.

There is a NixOS specific error that I'm having which I can see the log from `:LspLog` as below (because NixOS doesn't have `/lib/ld-linux-aarch64.so.1`).

```
291 [ERROR][2023-10-11 14:54:20] .../vim/lsp/rpc.lua:734> "rpc">"/home/user/.local/share/nvim/lazy/sg.nvim/dist/sg-lsp">"stderr"> "cannot execute /home/user/.local/share/nvim/lazy/sg.nvim/dist/sg-lsp: You are trying to run an unpatched binary on nixos, but you have not configured NIX_LD or NIX_LD_aarch64-linux. See https://github.com/Mic92/nix-ld for more details\n"
```

And I was able to run the binaries (`sg-lsp` and `sg-nvim-agent` via `nvim`) using https://github.com/Mic92/nix-ld on NixOS (although other binaries from other plugins started having issues due to `nix-ld` changing the environment).

Anyhow, if we disregard the specific case of on NixOS, I believe that you added a support for the arm/linux binaries 👏🏼!

For now, I believe the best bet for NixOS users would be building the binary themselves (which `sg.nvim` supports).

**Comment by ryuheechul** [comment]
just to add a bit more context on the nixos specific issue, the x86_64 binary from https://github.com/ryuheechul/termimagenator/releases/tag/v0.1.1 is also having the same problem on NixOS. (somehow arm one is working fine - I guess the difference is coming from how go build tool chains works vs rust one)

**Comment by tjdevries** [comment]
Ah, well we have some nix support for building the plugin, perhaps it could be modified to fix the problem? or it already works for nix users, they just have to be instructed to use that instead of the downloaded binaries?

**Comment by ryuheechul** [comment]
I can see that there is NixOS section in the readme.md and the example source code below.

```nix
sg = let
  system = "x86_64-linux";
  package = inputs.sg-nvim.packages.${system}.default;
in {
  inherit package;
  init = pkgs.writeTextFile {
    name = "sg.lua";
    text = ''
      return function()
        package.cpath = package.cpath .. ";" .. "${package}/lib/?.so"
      end
    '';
  };
};
```

However, I'm not sure what I need to do. Particularly with these points:
- what does `sg.lua` do and why do we need it?
- who is the "consumer" of `sg.lua`?
- I don't use nix flake to install packages (I use the legacy way which require `default.nix` from https://github.com/edolstra/flake-compat)

Just for a reference, previously with other plugins I just chose the crude and simple way to build them via:
- https://github.com/ryuheechul/dotfiles/blob/fba36522059885205f5016b9c981a30af7eece9b/nvim/lua/plugins/debug.lua#L31
- https://github.com/ryuheechul/dotfiles/blob/fba36522059885205f5016b9c981a30af7eece9b/nvim/shell/build-via-nix-shell.sh

---

I think I will stop reporting about the nixos specfic issue in this thread since:
- the original point was resolved 👍🏼 
- and I have no clear idea on how I can contribute more on helping with this nixos specific issue at the moment
